### DISCUSSION TITLE
Non buffered keyboard element

### DISCUSSION THREAD

**Discussion by me-nkr** [hneemann/Digital#1289] [Q&A]
So the keyboard component is a sequential one and buffers the input right?

Is there a way to use that component to create a component which outputs the current key being held or zero?

If not how can I have something like that?

Also having feature to have a custom key code map is also appreciated.

PS: I would like the component to give priority to key that was pressed last in case of multi key press, also here is the character set I want as of now

![image](https://github.com/hneemann/Digital/assets/60778845/a18eda26-8f9e-4816-b6bf-5c6d3ab27826)

**Comment by mengstr** [comment]
When I need something like that it's usually just a hex-keypad or something similar. Those are easy to "build yourself" with buttons and a priority encoder. If you set a label on the button then it can be mapped to your computer keyboard. It might be a bit tedious to do this with a full ascii keyboard though ;-) and you'll have a problem handling shift and control keys.

[keyb.dig.zip](https://github.com/hneemann/Digital/files/15209604/keyb.dig.zip)
<img width="650" alt="Screenshot 2024-05-04 at 14 02 34" src="https://github.com/hneemann/Digital/assets/325326/6c43c495-81f8-43e5-a827-e7fba5b70fe5">

**Reply by me-nkr** [reply]
The problem as stated is I can't type a lot of characters but and the keyboard layout I want is weird like some of them is ascii and some of them are just arbitrary also if there's a multi key press I want to switch selected key to the last pressed one. But this one was a eye opener for me though, like I never used the buttons component before and this is an awesome feature.

**Reply by mengstr** [reply]
Ok, try something like this.  The clock is needed for the monostable that is used to make the keyrepeat on the computer keyboard to be considered a single press and not a lot of repeated presses. The speed of the clock and the delay-value of the monostable should be adjusted to it is as short as possible without having repeated outputs instead of one continuous output value.

It's a really really ugly hack but I think it's the only way of doing this with the existing components.  

To have your own key mappings you can simply put a lookup-ROM at the output that you simply fill with values that converts the ascii values to whatever you want.  My Mac gives 0x11 for up-arrow. So if I want to make that 0x88 I'd simply put 0x88 in the 17'th (0x11) location of the ROM.
[keyb2.dig.zip](https://github.com/hneemann/Digital/files/15209842/keyb2.dig.zip)

<img width="772" alt="Screenshot 2024-05-04 at 15 24 49" src="https://github.com/hneemann/Digital/assets/325326/523e997b-c215-4f97-b728-e03daee02a2f">

**Reply by me-nkr** [reply]
out of topic how does the clock work like that? like the "wireless connection"?

**Reply by mengstr** [reply]
It just uses it. ;)  The monoflop is a really annoying component since `It can only be used if there is exactly one clock component present in the circuit. This clock component is used as time base to measure the time delay.` and most of my a bit more advanced designs uses multiple clocks. So I tend to avoid it

**Reply by me-nkr** [reply]
what I was asking is there is a clock that is not connected to anything and there is no tunnel, how does that work? like is it a bug, is it a feature?

**Reply by mengstr** [reply]
The monostable doesn't have a direct input for the clock. That is probably why it requires *exactly 1 clock* no more, no less, in the design.  It can then just look for *the* clock and use it internal in the software.  The blue wires are just a human-visible representation of the "netlist" that the software is using internally while processing.
 
I have no idea why it was made that way, maybe so it would be a bit more similar to a real physical monoflop. They don't have a clock input either, but rather sets the delay in an analogue way using a resistor and capacitor.  But since that is not possible here I guess it's a good middle way.

**Reply by hneemann** [reply]
That is mainly correct! The simulator knows no time. Only changes that take place one after the other or trigger each other. It is therefore not possible to simulate a component that triggers a change after a certain time - set with an R-C circuit. If you want to simulate a monoflop reproducibly, the time must come from somewhere. And the clock component in the circuit is used for this purpose. The delay time of the monoflop can then be determined in units of the clock period.

**Comment by hneemann** [comment]
What should happen if several buttons are pressed at the same time? Keyboards are normally connected in a matrix so that each key can be queried to determine whether it is currently pressed or not.

**Reply by me-nkr** [reply]
well in the case of nand2tetris, it's a bit ambiguous. But here's what I have made out

- there is no concept of multi key press
- when shift is pressed the alternate character of the key is used, but still there is no concept called shift, afaik
- when two keys are pressed together the last pressed key is chosen, if the last pressed key is shift, then shift + last pressed key to shift is chosen

**Reply by hneemann** [reply]
But is not everything in the Nand2Tetris project implemented by yourself? Then you can connect the keyboard however you like. And a computer game like Tetris can easily be implemented with the existing keyboard.
And if you need a control pad in which several buttons can be queried simultaneously, you can use the existing buttons.

**Reply by me-nkr** [reply]
In nand2tetris the scan code of button you press becomes the value of a memory address. once you release it the value resets to 0. For programs the value is queried every clock cycle. And I am trying to make the computer I make as true to the  course as possible. Please feel free to mention anything I missed...
